### PR TITLE
Semantics attribute definition file

### PR DESCRIPTION
**Pull request by zoecarver**
This patch creates a definition file for all non-specialized semantics attributes. Its the first in what will probably be three patches. The goal is to enforce that all semantics attributes are constant string literals that are defined in one place which will reduce the possibility for error and create a centralized place for all semantics attributes to live.

The next patch will add a category so that we can replace code such as the following: 
```swift
hasSemanticsAttrThatStartsWith("array.")
```
with something like
```swift
hasSemanticsAttrCategory(ARRAY)
```

Last, it might be nice to have some way to enforce that `hasSemanticsAttr` is passed a defined attribute. 

Additionally, here are some semantics that I found references to in the compiler but could not find references to in the standard library:
```
verify.ownership.sil.never
verify.ownership.sil.never
self_no_escaping_closure
pair_no_escaping_closure
inline_late
```

### CODE REVIEW

**Comment by zoecarver** [comment]
cc @gottesmm

File: include/swift/Basic/Semantics.def [hunk 1]
```diff
@@ -0,0 +1,78 @@
+//===--- Semantics.def - Semantics Attribute Definitions -------------*- C++ -*-===//
```
Comment (owenv) [inline include/swift/Basic/Semantics.def]: This (and the corresponding header) should probably go in `AST/` as opposed to `Basic` so it's in the same lib as `hasSemanticsAttr`

`Semantics.def` Also feels a bit vague to me. What do you think about `SemanticAttrs.def`?

File: include/swift/Basic/Semantics.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SEMANTICS_H
+#define SWIFT_SEMANTICS_H
+
+#include "llvm/ADT/StringRef.h"
+
+namespace swift {
+#define SEMA_ATTR(NAME, C_STR) constexpr static const StringLiteral NAME = #C_STR;
```
Comment (owenv) [inline include/swift/Basic/Semantics.h]: These should probably go in their own namespace so they don't clutter the top level swift namespace, similar to how diagnostic definitions are in `swift::diag::`

File: include/swift/Basic/Semantics.def [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
```
Comment (owenv) [inline include/swift/Basic/Semantics.def]: nit: copyright year here and in Semantics.h

File: include/swift/Basic/Semantics.def [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This macro is used to map the global definition name of a semantic attribute to its raw value.
```
Comment (owenv) [inline include/swift/Basic/Semantics.def]: nit: wrap to 80 chars. See also a few of the definitions below and the header of Semantics.h.

File: include/swift/Basic/Semantics.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Implementation of the matching definition file.
+// This file holds all semantics attributes as consant stirng literals.
```
Comment (owenv) [inline include/swift/Basic/Semantics.h]: Small typo: `constant string`

**Review by owenv — COMMENTED** [review]
I think this is a reasonable refactor, but someone more familiar with `@_semantics` should take a look as well. In the meantime I just have a few small comments.

File: include/swift/Basic/Semantics.h (hunk 2 above)
Comment (zoecarver) [reply include/swift/Basic/Semantics.h]: I'm not opposed to this but let's see what others think.

File: include/swift/AST/SemanticAttrs.def [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This macro is used to map the global definition name of a semantic
```
Comment (gottesmm) [inline include/swift/AST/SemanticAttrs.def]: Please follow LLVM style here:

https://llvm.org/docs/CodingStandards.html#file-headers

Notice how this part of the file header has a \file and also has 3 slashes.

File: include/swift/AST/SemanticAttrs.def [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+// attribute to its raw value.
+//   NAME: the global name used in the compiler
+//   C_STR: the raw value used in swift
+//
+// SEMA_ATTR(NAME, C_STR)
+//
+//===----------------------------------------------------------------------===//
+
+SEMA_ATTR(STRING_EQUALS, "string.equals")
```
Comment (gottesmm) [inline include/swift/AST/SemanticAttrs.def]: Why not use SEMANTICS_ATTR? Then it is clear that you are not referring to sema. But maybe I am being too worried.

Comment (gottesmm) [reply include/swift/AST/SemanticAttrs.def]: Also, I would suggest maybe adding a check with a nice error if SEMA_ATTR is not defined and you included the function.

**Review by gottesmm — COMMENTED** [review]
I am also tending to think we should put this into a namespace like semantics or perhaps strings? Not sure. @atrick @jrose-apple any ideas?

**Comment by zoecarver** [comment]
I added a `semantics` namespace. Now that it's there, it will be easy to change or remove with a find/replace all.

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
This looks good to me. Since nobody else has commented on this in 2 weeks, I think we can merge it if the tests pass.

**Comment by slavapestov** [comment]
If you want, you can remove the semantics not used by the stdlib in a subsequent PR.

**Comment by slavapestov** [comment]
Looks like a few test failures snuck in since this was last tested:

```
01:08:48 Failing Tests (9):
01:08:48     Swift(linux-x86_64) :: SILOptimizer/string_switch.swift
01:08:48     Swift(linux-x86_64) :: SILOptimizer/no_size_specialization.swift
01:08:48     Swift(linux-x86_64) :: SILOptimizer/pound_assert.swift
01:08:48     Swift(linux-x86_64) :: SILOptimizer/diagnostic_constant_propagation.swift
01:08:48     Swift(linux-x86_64) :: SILOptimizer/constant_evaluator_test.sil
01:08:48     Swift(linux-x86_64) :: SILOptimizer/concat_string_literals.64.swift
01:08:48     Swift(linux-x86_64) :: SILOptimizer/constant_evaluable_subset_test.swift
01:08:48     Swift(linux-x86_64) :: SILOptimizer/constant_propagation.sil
01:08:48     Swift(linux-x86_64) :: SILOptimizer/OSLogPrototypeCompileTest.sil
```

Hopefully it's a simple typo somewhere.

**Comment by zoecarver** [comment]
Ha! I'm surprised those are the only ones that failed. I accidentally initialized all the semantics as `foo = "\"abc\""` not `foo = "abc"`. Should be fixed now.

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by zoecarver** [comment]
@slavapestov looks like the bots maybe didn't get triggered?

**Comment by theblixguy** [comment]
Yeah, there was an [outage](https://forums.swift.org/t/unexpected-swift-ci-outage-11-19/30949), which is now fixed.

@swift-ci please test

**Comment by zoecarver** [comment]
Thank you @slavapestov!

**Pull request merged**